### PR TITLE
Chore: Stimulus controller rewrite for user signup form

### LINKED ISSUE #2377: Update Javascript handling the New User Form Validations
#### Description
Some of the legacy Javascript can now be updated so we can eventually remove the depedency on jquery.

As part of this we need to update [this file](https://github.com/TheOdinProject/theodinproject/blob/main/app/javascript/src/js/formValidations.js) to use a Stimulus Controller.

Suggest reading to help you: https://stimulus.hotwired.dev/ and https://www.betterstimulus.com/

#### Acceptance Criteria:

1. The Form still validates as it currently does
2. A stimulus controller is created to handle the interaction

#### Additional Information:

Stimulus controllers are all about finding the correct level of abstraction. The suggested reading for better stimulus covers this.

### PR DESCRIPTION
**Pull request by bscottnz**
-   [x] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/theodinproject/blob/main/CONTRIBUTING.md)
-   [x] The title of this PR follows the `keyword: brief description of change` format, using one of the following keywords:
  - `Feature` - adds new or amends existing user-facing behaviour
  - `Chore` - changes that have no user-facing value, refactors, dependency bumps, etc
  - `Fix` - bug fixes

Complete the following checkbox ONLY IF it is applicable to your PR. You can complete it later if it is not currently applicable:
-   [x] I have verified all tests and linters pass against my changes, and/or I have included automated tests where applicable

<hr>

**1. Because:**
Closes #2377

**2. This PR:**
rewrite new user signup validation to use stimulus controller.

**3. Additional Information:**
I had to add `data: { disable_with: false } ` to the new user submit button as the button would disable during failed validation - https://stackoverflow.com/questions/42016113/rails-data-disable-with-re-enabling-button

When signing in with a blank email & password input, it will now validate client side instead of submitting and getting the wrong username & password banner. I had to remove one test to reflect this change in behavior. (it would look for the banner)

### CODE REVIEW

File: app/javascript/src/js/formValidations.js [hunk 1]
```diff
@@ -1,121 +1,39 @@
...
 }
 
-function createErrorDiv(input, inputParent) {
-  const divName = document.createElement('div');
-  divName.classList.add('form__error-message', 'push-down', input.id);
-  inputParent.parentNode.insertBefore(divName, inputParent.nextSibling);
-  return divName;
-}
+const application = Application.start();
+application.register('signup-validation', SignupValidation)
```
Comment (KevinMulhern) [inline app/javascript/src/js/formValidations.js:39]: Stimulus controllers conventionally live in the `app/javascript/controllers` directory. Can we move this file to that directory and rename it to `form_validation_controller.js` please?

Then we should be able to remove these lines for registering the controller. That will be automatically done for us in the `controller/index.js` file.

File: app/javascript/src/js/formValidations.js [hunk 2]
```diff
@@ -1,121 +1,39 @@
...
-      message: '^The passwords do not match',
-    },
-  },
-  'user[current_password]': {
-    length: {
-      minimum: 6,
-      maximum: 128,
-    },
-  },
-  'user[learning_goal]': {
```
Comment (KevinMulhern) [inline app/javascript/src/js/formValidations.js:33]: The old validations must have handled the settings page validations too. Can we add a validation for the learning goal and apply validations to the [profile](https://github.com/TheOdinProject/theodinproject/blob/main/app/views/devise/registrations/_profile.html.erb), [learning goal](https://github.com/TheOdinProject/theodinproject/blob/main/app/views/devise/registrations/_learning_goal.html.erb) and [password](https://github.com/TheOdinProject/theodinproject/blob/main/app/views/devise/registrations/_password.html.erb) settings page forms too please?

File: app/javascript/src/js/formValidations.js [hunk 3]
```diff
@@ -1,121 +1,39 @@
...
-    length: {
+import { Application } from "stimulus";
+import { ValidationController } from "stimulus-validation"
+// https://github.com/jwald1/stimulus-validation
+
+export default class SignupValidation extends ValidationController {
```
Comment (KevinMulhern) [inline app/javascript/src/js/formValidations.js:7]: just to make this more generic and reusable can it be renamed to:
```suggestion
export default class FormValidation extends ValidationController {
```

File: app/javascript/src/js/formValidations.js [hunk 4]
```diff
@@ -1,121 +1,39 @@
...
-    length: {
-      minimum: 6,
-      maximum: 128,
-    },
-  },
-  'user[learning_goal]': {
-    length: {
+import { Application } from "stimulus";
+import { ValidationController } from "stimulus-validation"
+// https://github.com/jwald1/stimulus-validation
```
Comment (KevinMulhern) [inline app/javascript/src/js/formValidations.js:5]: Can this comment be removed please?

File: app/views/devise/registrations/_form.html.erb [hunk 5]
```diff
@@ -1,44 +1,44 @@
...
-
-    <div class="form__section">
-      <span class="fas fa-user form__icon"></span>
-      <%= form.text_field :username, required: true, autofocus: true, placeholder: 'Username', class: 'form__element form__element--with-icon', data: {test_id: 'username_field'} %>
+  <%= form_for(resource, as: resource_name, url: registration_path(resource_name), html: { class: 'form', 'data-controller':"signup-validation", 'data-action':"submit->signup-validation#validateAll" }) do |form| %>
+
+    <div class="form-field mb-4">
+      <div class="form__section mb-0">
+        <span class="fas fa-user form__icon"></span>
+        <%= form.text_field :username, 'data-action':"blur->signup-validation#validate", 'data-attr':"signup-validation.username", required: true, placeholder: 'Username', class: 'form__element form__element--with-icon', data: {test_id: 'username_field'} %>
```
Comment (KevinMulhern) [inline app/views/devise/registrations/_form.html.erb:7]: We can add any data attributes to the ruby data hash to cut down on a little bit of code

```suggestion
        <%= form.text_field(
          :username, 
          required: true, 
          placeholder: 'Username', 
          class: 'form__element form__element--with-icon', 
          data: { test_id: 'username_field', action: "blur->signup-validation#validate", attr: "signup-validation.username"} 
        )%>
```

Can we do this for all the form fields with validation data attributes?

File: app/javascript/src/js/formValidations.js [hunk 6]
```diff
@@ -1,121 +1,39 @@
 /* eslint-disable */
```
Comment (KevinMulhern) [inline app/javascript/src/js/formValidations.js:1]: We should hopefully be able to re-enable eslint after this clean up 💪

File: spec/system/user_login_spec.rb [hunk 7]
```diff
@@ -19,18 +19,6 @@
       end
     end
 
-    context 'with blank input fields' do
```
Comment (KevinMulhern) [inline spec/system/user_login_spec.rb:22]: Nice!

**Review by KevinMulhern — CHANGES_REQUESTED** [review]
Looking good 💪  most of the feedback it is just Rails convention stuff.

File: app/javascript/src/js/formValidations.js (hunk 1 above)
Comment (bscottnz) [reply app/javascript/src/js/formValidations.js:39]: this no longer works when removing these 2 lines. it does bring it in as shown here

image

image

but the validation no longer works. something to do with the stimulus-validation package maybe?

File: app/javascript/src/js/formValidations.js (hunk 6 above)
Comment (bscottnz) [reply app/javascript/src/js/formValidations.js:1]: I had this error `import/no-extraneous-dependencies` with this line `import { Application } from 'stimulus';` so i have kept eslint disable for this one line only.

following the linter instructions to npm install introduced other errors, as did yarn add. so i removed both of these installs and disabled linting on that line instead.

**Comment by bscottnz** [comment]
Thanks for the feedback, I have fixed up these issues. I also added a new rspec test for the new validation on the profile settings page for good measure.

File: app/javascript/src/js/formValidations.js (hunk 1 above)
Comment (KevinMulhern) [reply app/javascript/src/js/formValidations.js]: Yeah I just noticed the library is a little out of date. It looks like it was built for Stimulus 2.0 but we're on 3. My fault, for not looking into that library enough before suggesting it  😅

But it's working and is so much better and cleaner than what we had before. Let's get this in and revisit somewhere down the line.

**Review by KevinMulhern — APPROVED** [review]
Looks great @bscottnz, just a merge conflict to fix and we're good to go 🚀

**Comment by bscottnz** [comment]
im having trouble rebasing the yarn.lock merge conflict commit, is this something that gets merged behind the scenes

**Pull request closed without merging**